helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -501,21 +512,64 @@ private void calculateAndUpdateBaseline(ClusterModel clusterModel, RebalanceAlgo
     _baselineCalcLatency.endMeasuringLatency();
     LOG.info("Global baseline calculation completed and has been persisted into metadata store.");
 
-    if (isBaselineChanged && shouldSchedulePartialRebalance) {
+    if (isBaselineChanged && shouldTriggerMainPipeline) {
       LOG.info("Schedule a new rebalance after the new baseline calculation has finished.");
-      RebalanceUtil.scheduleOnDemandPipeline(clusterName, 0L, false);
+      RebalanceUtil.scheduleOnDemandPipeline(clusterData.getClusterName(), 0L, false);
     }
   }
 
-  private Map<String, ResourceAssignment> partialRebalance(
+  private void partialRebalance(
       ResourceControllerDataProvider clusterData, Map<String, Resource> resourceMap,
       Set<String> activeNodes, final CurrentStateOutput currentStateOutput,
       RebalanceAlgorithm algorithm)
       throws HelixRebalanceException {
+    // If partial rebalance is async and the previous result is not completed yet,
+    // do not start another partial rebalance.
+    if (_asyncPartialRebalanceEnabled && _asyncPartialRebalanceResult != null
+        && !_asyncPartialRebalanceResult.isDone()) {
+      return;
```

Comment:
For my own understanding, does this mean we have more frequent emergency rebalance? 

also nit: Let's log something here.

Reply:
More frequent emergency rebalance? Do you mean the sync pipeline is triggered more often? I don't fully understand the question. 

I will add more logs in a followup PR. :)

Reply:
Yes, I'm trying to understand what it means if we enter this block. Essentially, if the sync pipeline is triggered too frequently, we'll see more attempts to start partial rebalance while there is one still in progress. Is that right?

Reply:
That's correct, but that also wouldn't actually spawn a thread.